Source/WebCore/html/HTMLCollection.cpp

```diff
@@ -86,13 +86,13 @@ static NodeListInvalidationType invalidationTypeExcludingIdAndNameAttributes(Col
     case TableRows:
     case TRCells:
     case SelectOptions:
+    case DataListOptions:
```

Comment:
Can you add a test for the DataListOptions part of the change?

Reply:
@geoffreygaren I would love to but I don't know how that test should look like, like what attribute should change to ensure that collection is correct given the collection is merely a `getElementsByTagName("option")` 🤔

Reply:
I think the code says that DataListOptions used to invalidate on any attribute change, and now doesn't have to. If so, I would write a test that made a DataListOptions, then did an attribute change, then made a DataListOptions again, and verified that the two DataListOptions objects were equal. But maybe I've misunderstood the behavior.

Reply:
I think the confusion comes from the fact that the caching we're talking about here is a caching of the list "items" on DOM side. This doesn't have anything to do with JS wrappers caching.

For performance reason, when you iterate over a HTMLCollection or LiveNodeList for the first time, we cache the nodes that belong to the list as we traverse the DOM tree. This way, on next iteration, we don't need to traverse the DOM tree anymore.

The risk from this change is not JS object identify but rather, iterating over a HTMLCollection or LiveNodeList and getting outdated nodes (nodes that should no longer be part of the list) or missing nodes (that should now be part of the list), because we failed to invalidate our list cache on DOM tree mutation.

Reply:
So I agree with Alexey that there should be no web-facing behavior change, unless there is a bug.

Reply:
@geoffreygaren The modified function, `invalidationTypeExcludingIdAndNameAttributes`, invalidates only the internal cache of `DataListOptions`, and can't possibly affect the object identity in JS.

Object identity is maintained by returning a `Ref` from

```cpp
Ref<HTMLCollection> HTMLDataListElement::options()
{
    return ensureRareData().ensureNodeLists().addCachedCollection<GenericCachedHTMLCollection<CollectionTypeTraits<DataListOptions>::traversalType>>(*this, DataListOptions);
}
```

and then wrapping it in `JSDOMWrapper`, as well as saving it to `NodeListsNodeData`. It's getting cleared only if JS loses reference to returned `dataList.options` object, independent of cache invalidation.

With that said, I would rather not add a test along with this change to avoid confusing future `git log` readers 😃

Reply:
Got it! I misunderstood.